tls/datadog_checks/tls/tls.py

```diff
@@ -0,0 +1,353 @@
+# (C) Datadog, Inc. 2019
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import socket
+import ssl
+from datetime import datetime
+from os.path import expanduser, isdir
+
+import service_identity
+from cryptography.hazmat.backends import default_backend
+from cryptography.x509 import load_der_x509_certificate, load_pem_x509_certificate
+from six import text_type
+from six.moves.urllib.parse import urlparse
+
+from datadog_checks.base import AgentCheck, ConfigurationError, is_affirmative
+
+from .utils import closing, days_to_seconds, get_protocol_versions, is_ip_address, seconds_to_days
+
+# Python 3 only
+PROTOCOL_TLS_CLIENT = getattr(ssl, 'PROTOCOL_TLS_CLIENT', ssl.PROTOCOL_TLS)
+
+
+class TLSCheck(AgentCheck):
+    SERVICE_CHECK_CAN_CONNECT = 'tls.can_connect'
+    SERVICE_CHECK_VERSION = 'tls.version'
+    SERVICE_CHECK_VALIDATION = 'tls.cert_validation'
+    SERVICE_CHECK_EXPIRATION = 'tls.cert_expiration'
+
+    DEFAULT_EXPIRE_DAYS_WARNING = 14
+    DEFAULT_EXPIRE_DAYS_CRITICAL = 7
+    DEFAULT_EXPIRE_SECONDS_WARNING = days_to_seconds(DEFAULT_EXPIRE_DAYS_WARNING)
+    DEFAULT_EXPIRE_SECONDS_CRITICAL = days_to_seconds(DEFAULT_EXPIRE_DAYS_CRITICAL)
+
+    def __init__(self, name, init_config, instances):
+        super(TLSCheck, self).__init__(name, init_config, instances)
+
+        self._name = self.instance.get('name')
+        self._local_cert_path = self.instance.get('local_cert_path', '')
+        self._timeout = float(self.instance.get('timeout', 10))
```

Comment:
More curiosity, but why underscore for some of these?

Reply:
So we don't ever clash with something the base class defines

Reply:
I see, makes sense.